crates/bitwarden-crypto/src/service/context.rs

```diff
@@ -0,0 +1,357 @@
+use std::sync::{RwLockReadGuard, RwLockWriteGuard};
+
+use rsa::Oaep;
+use zeroize::Zeroizing;
+
+use super::Keys;
+use crate::{
+    derive_shareable_key,
+    service::{key_store::KeyStore, AsymmetricKeyRef, SymmetricKeyRef},
+    AsymmetricCryptoKey, AsymmetricEncString, CryptoError, EncString, Result, SymmetricCryptoKey,
+};
+
+// This is to abstract over the read-only and read-write access to the global keys
+// inside the CryptoServiceContext. The read-write access should only be used internally
+// in this crate to avoid users leaving the crypto store in an inconsistent state,
+// but for the moment we have some operations that require access to it.
+pub trait GlobalAccessMode<'a, SymmKeyRef: SymmetricKeyRef, AsymmKeyRef: AsymmetricKeyRef> {
```

Comment:
**question**: Is there a purpose in making the types generic?

Reply:
So initially, the way I thought about the `CryptoContext` was that it would have a read-only view of the global keys, and a read-write exclusive store to deal with intermediate and ephemeral keys, like cipher keys.

The reason to enforce the read-only view was that it would allow you to parallelize decryption by just creating a bunch of independent contexts. This works great for decryptable/encryptable implementations, which don't have to modify the global keys ever.

What I didn't initially consider was that this context concept would also be useful when initializing the global keys themselves. For example, the context provides utilities for decrypting the org keys directly into the key store, but of course, as-is, the context can't store things into the global store. 

So to avoid duplicating the context code, I made the `Context` generic over the read/read-write status over the global store. This is a bit awkward and leads to strange things like the always-failing `get_mut` you mention here: https://github.com/bitwarden/sdk-internal/pull/7#discussion_r1824615486. I'm now thinking it would have been simpler with an `enum GlobalAccessMode { ReadOnly(RWLockReadGuard), ReadWrite(RWLockWriteGuard) }`

I think this change, combined with the local key refs not being a separate type and instead being a runtime check, makes things a bit weird. I want to look into improving the story around local keys but that's another thing.